File: EVChargingBackend/Models/Notification.cs
/*
 * File: Notification.cs
 * Project: EV Charging Station Booking System
 * Description: Notification model for system notifications
 * Author: EV Charging System
 * Date: October 7, 2025
 */

using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace EVChargingBackend.Models
{
    /// <summary>
    /// Represents a notification in the system
    /// </summary>
    public class Notification
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string? Id { get; set; }

        [BsonElement("recipientNIC")]
        public string RecipientNIC { get; set; } = string.Empty;

        [BsonElement("title")]
        public string Title { get; set; } = string.Empty;

        [BsonElement("message")]
        public string Message { get; set; } = string.Empty;

        [BsonElement("type")]
        public NotificationType Type { get; set; }

        [BsonElement("relatedEntityId")]
        public string? RelatedEntityId { get; set; }

        [BsonElement("relatedEntityType")]
        public string? RelatedEntityType { get; set; }

        [BsonElement("isRead")]
        public bool IsRead { get; set; } = false;

        [BsonElement("isDelivered")]
        public bool IsDelivered { get; set; } = false;

        [BsonElement("priority")]
        public NotificationPriority Priority { get; set; } = NotificationPriority.Normal;

        [BsonElement("createdAt")]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        [BsonElement("readAt")]
        public DateTime? ReadAt { get; set; }

        [BsonElement("deliveredAt")]
        public DateTime? DeliveredAt { get; set; }

        [BsonElement("expiresAt")]
        public DateTime? ExpiresAt { get; set; }

        [BsonElement("metadata")]
        public Dictionary<string, object>? Metadata { get; set; }
    }

    /// <summary>
    /// Types of notifications in the system
    /// </summary>
    public enum NotificationType
    {
        BookingConfirmation = 0,
        BookingCancellation = 1,
        BookingReminder = 2,
        StationUpdate = 3,
        SystemAlert = 4,
        PaymentConfirmation = 5,
        BookingExpired = 6
    }

    /// <summary>
    /// Priority levels for notifications
    /// </summary>
    public enum NotificationPriority
    {
        Low = 0,
        Normal = 1,
        High = 2,
        Critical = 3
    }
}

File: EVChargingMobile/app/src/main/java/com/example/evcharging/models/Notification.java
package com.example.evcharging.models;

import com.google.gson.annotations.SerializedName;
import java.util.Date;
import java.util.Map;

// This class now accurately reflects the backend's Notification.cs model
public class Notification {

    @SerializedName("id")
    public String id;

    @SerializedName("recipientNIC")
    public String recipientNIC;

    @SerializedName("title")
    public String title;

    @SerializedName("message")
    public String message;

    // Backend sends 'type' as a number (enum)
    @SerializedName("type")
    public int type;

    @SerializedName("relatedEntityId")
    public String relatedEntityId;

    @SerializedName("isRead")
    public boolean isRead;

    // Backend sends 'priority' as a number (enum)
    @SerializedName("priority")
    public int priority;

    @SerializedName("createdAt")
    public Date createdAt;

    // Optional fields
    @SerializedName("metadata")
    public Map<String, Object> metadata;

    // Default constructor for Gson
    public Notification() {}
}
